model/DataObject.php

```diff
@@ -1881,168 +1940,203 @@ public function has_many($component = null, $classOnly = true) {
 	}
 
 	/**
+	 * Return data for a specific has_many component.
+	 * @param string $component
+	 * @param bool $classOnly If this is TRUE, than any has_many relationships in the form "ClassName.Field" will have
+	 *        the field data stripped off. It defaults to TRUE.
+	 * @return string|false
+	 */
+	public function hasManyComponent($component, $classOnly = true) {
+		$hasMany = (array)Config::inst()->get($this->class, 'has_many', Config::INHERITED);
+
+		if($hasMany && array_key_exists($component, $hasMany)) {
+			$hasMany = $hasMany[$component];
+		} else {
+			return false;
+		}
+
+		return ($classOnly) ? preg_replace('/(.+)?\..+/', '$1', $hasMany) : $hasMany;
+	}
+
+	/**
+	 * @deprecated 4.0 Method has been renamed to manyManyExtraFields()
+	 * @param string $component
+	 * @return array
+	 */
+	public function many_many_extraFields($component = null) {
+		Deprecation::notice('4.0', 'Please use manyManyExtraFields() instead');
+		return $this->manyManyExtraFields($component);
+	}
+
+	/**
 	 * Return the many-to-many extra fields specification.
 	 *
 	 * If you don't specify a component name, it returns all
 	 * extra fields for all components available.
 	 *
-	 * @param string $component Name of component
-	 * @return array
+	 * @param string $component Deprecated - Name of component
+	 * @return array|null
 	 */
-	public function many_many_extraFields($component = null) {
-		$classes = ClassInfo::ancestry($this);
-
-		foreach($classes as $class) {
-			if(in_array($class, array('ViewableData', 'Object', 'DataObject'))) continue;
-			$relationName = null;
-
-			// Find extra fields for one component
-			if($component) {
-				$SNG_class = singleton($class);
-				$extraFields = $SNG_class->stat('many_many_extraFields');
-
-				// Extra fields are immediately available on this class
-				if(isset($extraFields[$component])) {
-					return $extraFields[$component];
-				}
-
-				$manyMany = $SNG_class->stat('many_many');
-				$candidate = (isset($manyMany[$component])) ? $manyMany[$component] : null;
-				if($candidate) {
-					$SNG_candidate = singleton($candidate);
-					$candidateManyMany = $SNG_candidate->stat('belongs_many_many');
-
-					// Find the relation given the class
-					if($candidateManyMany) foreach($candidateManyMany as $relation => $relatedClass) {
-						if($relatedClass == $class) {
-							$relationName = $relation;
-							break;
-						}
-					}
-
-					if($relationName) {
-						$extraFields = $SNG_candidate->stat('many_many_extraFields');
-						if(isset($extraFields[$relationName])) {
-							return $extraFields[$relationName];
-						}
-					}
-				}
```

Comment:
The best I can do is say:

1. There’s no documentation that suggests you can define extra fields on the `belongs_many_many` side (mainly because there’s almost no documentation on extra fields at all!)
2. The [code responsible](https://github.com/silverstripe/silverstripe-framework/blame/3a7e24a2208f58faa7ae8f1b5e575d39a65c918a/model/DataObject.php#L3096) for fetching extra fields to build the join table hasn’t changed since 2007 - it only checks for `many_many_extraFields` if it encounters `many_many` - it doesn’t even look at `belongs_many_many`.